From 08bb1d9a06ca3b8dd2216e1fc9f5608227790707 Mon Sep 17 00:00:00 2001
From: Kai Sommerfeld <kai.sommerfeld@gmx.com>
Date: Fri, 12 Jun 2020 18:03:14 +0200
Subject: [PATCH 1/2] [PVR] CGUIEPGGridContainer: Update layout early in GUI
 thread, not in timeline refresh thread, for performance and thread safety
 reasons (fix a race).

---
 xbmc/pvr/guilib/GUIEPGGridContainer.cpp | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/xbmc/pvr/guilib/GUIEPGGridContainer.cpp b/xbmc/pvr/guilib/GUIEPGGridContainer.cpp
index 5b8ca6192f6c..da36405002af 100644
--- a/xbmc/pvr/guilib/GUIEPGGridContainer.cpp
+++ b/xbmc/pvr/guilib/GUIEPGGridContainer.cpp
@@ -1643,6 +1643,8 @@ void CGUIEPGGridContainer::LoadLayout(TiXmlElement* layout)
     m_rulerLayouts.back().LoadLayout(itemElement, GetParentID(), false, m_width, m_height);
     itemElement = itemElement->NextSiblingElement("rulerlayout");
   }
+
+  UpdateLayout();
 }
 
 std::string CGUIEPGGridContainer::GetDescription() const
@@ -1774,7 +1776,6 @@ void CGUIEPGGridContainer::SetTimelineItems(const std::unique_ptr<CFileItemList>
   {
     CSingleLock lock(m_critSection);
 
-    UpdateLayout();
     iRulerUnit = m_rulerUnit;
     iFirstChannel = m_channelOffset;
     iChannelsPerPage = m_channelsPerPage;

From 1e42320be1b40c33619a5bb526cbf3732e130c15 Mon Sep 17 00:00:00 2001
From: Kai Sommerfeld <kai.sommerfeld@gmx.com>
Date: Fri, 12 Jun 2020 18:13:49 +0200
Subject: [PATCH 2/2] [PVR] Make CGUIEPGGridContainer::UpdateLayout
 thread-safe; some of the modified members will be read by timeline refresh
 thread.

---
 xbmc/pvr/guilib/GUIEPGGridContainer.cpp | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/xbmc/pvr/guilib/GUIEPGGridContainer.cpp b/xbmc/pvr/guilib/GUIEPGGridContainer.cpp
index da36405002af..be25a4a9b813 100644
--- a/xbmc/pvr/guilib/GUIEPGGridContainer.cpp
+++ b/xbmc/pvr/guilib/GUIEPGGridContainer.cpp
@@ -1862,6 +1862,8 @@ void CGUIEPGGridContainer::UpdateLayout()
       oldRulerLayout == m_rulerLayout && oldRulerDateLayout == m_rulerDateLayout)
     return; // nothing has changed, so don't update stuff
 
+  CSingleLock lock(m_critSection);
+
   m_channelHeight = m_channelLayout->Size(VERTICAL);
   m_channelWidth = m_channelLayout->Size(HORIZONTAL);
 
